### PR TITLE
Expose google provisioning model to be able to acquire spot instances

### PR DESCRIPTION
**Pull request by tlamr**
Signed-off-by: Tomas Lamr <tomas@lamr.org>

**What this PR does / why we need it**:
Hello, we wanted to expose node selection API from google (SPOT/STANDARD) so that we can create a node pool of spot instances as in was.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Optional Release Note**:

```release-note
Support for specifying provisioning model for instances on GCE
```

### CODE REVIEW

**Comment by tlamr** [comment]
Difference between preemptible and spot according to them:

However, Spot VMs provide new features that are not supported for preemptible VMs. For example, preemptible VMs can only run for up to 24 hours at a time, but Spot VMs do not have a maximum runtime. This page describes preemptible virtual machine (VM) instances.

**Comment by ahmedwaleedmalik** [comment]
@tlamr can you please look into the failing CI jobs?

**Comment by tlamr** [comment]
Hello @ahmedwaleedmalik, I'm not an expert but I can see 20x this error and then it seems to give up.
`Error: network limit reached (resource_limit_exceeded)

  with hcloud_network.net,
  on hetzner.tf line 10, in resource "hcloud_network" "net":
  10: resource "hcloud_network" "net" {
...
Creating cloud provider env failed!
Cleaning up machines.`

**Comment by ahmedwaleedmalik** [comment]
> Error: network limit reached (resource_limit_exceeded)

This is unrelated to your PR. I'll take care of this issue. 

Meanwhile can you check the logs for https://public-prow.loodse.com/view/gs/prow-dev-public-data/pr-logs/pull/kubermatic_machine-controller/1252/pull-machine-controller-lint/1514635654829445120

File: pkg/cloudprovider/provider/gce/types/types.go [hunk 1]
```diff
@@ -38,6 +38,8 @@ type CloudProviderSpec struct {
 	Network               providerconfigtypes.ConfigVarString `json:"network"`
 	Subnetwork            providerconfigtypes.ConfigVarString `json:"subnetwork"`
 	Preemptible           providerconfigtypes.ConfigVarBool   `json:"preemptible"`
+	AutomaticRestart      providerconfigtypes.ConfigVarBool   `json:"automaticRestart,omitempty"`
+	ProvisioningModel     providerconfigtypes.ConfigVarString `json:"provisioningModel,omitempty"`
```
Comment (ahmedwaleedmalik) [inline pkg/cloudprovider/provider/gce/types/types.go:42]:
```suggestion
	AutomaticRestart      *providerconfigtypes.ConfigVarBool   `json:"automaticRestart,omitempty"`
	ProvisioningModel     *providerconfigtypes.ConfigVarString `json:"provisioningModel,omitempty"`
```

These should be optional values

**Review by ahmedwaleedmalik — CHANGES_REQUESTED** [review]

**Comment by ahmedwaleedmalik** [comment]
@tlamr I have updated the description of your PR and made it compliant with the required template.

**Comment by tlamr** [comment]
@ahmedwaleedmalik Awesome, thank you! :)

File: pkg/cloudprovider/provider/gce/provider.go [hunk 2]
```diff
@@ -265,7 +265,9 @@ func (p *Provider) Create(machine *clusterv1alpha1.Machine, data *cloudprovidert
 		Disks:             disks,
 		Labels:            labels,
 		Scheduling: &compute.Scheduling{
-			Preemptible: cfg.preemptible,
+			Preemptible:       cfg.preemptible,
+			AutomaticRestart:  &cfg.automaticRestart,
+			ProvisioningModel: cfg.provisioningModel,
```
Comment (ahmedwaleedmalik) [inline pkg/cloudprovider/provider/gce/provider.go:270]: Need to check for nil before assigning these values.

Comment (ahmedwaleedmalik) [inline pkg/cloudprovider/provider/gce/provider.go:270]: Check for nil before assigning these values

File: pkg/cloudprovider/provider/gce/config.go [hunk 3]
```diff
@@ -101,6 +101,8 @@ type config struct {
 	network               string
 	subnetwork            string
 	preemptible           bool
+	automaticRestart      bool
+	provisioningModel     string
```
Comment (ahmedwaleedmalik) [inline pkg/cloudprovider/provider/gce/config.go:105]: Preferably you should use pointers here

File: pkg/cloudprovider/provider/gce/provider.go [hunk 4]
```diff
@@ -265,7 +265,9 @@ func (p *Provider) Create(machine *clusterv1alpha1.Machine, data *cloudprovidert
 		Disks:             disks,
 		Labels:            labels,
 		Scheduling: &compute.Scheduling{
-			Preemptible: cfg.preemptible,
+			Preemptible:       cfg.preemptible,
+			AutomaticRestart:  &cfg.automaticRestart,
```
Comment (ahmedwaleedmalik) [inline pkg/cloudprovider/provider/gce/provider.go:269]: This is an incomplete implementation as you are not consuming values from `ProviderSpec` for the new fields i.e. AutomaticRestart and ProvisioningModel.

Please go through the handling of `preemptible` in `newConfig` as an example:

https://github.com/kubermatic/machine-controller/blob/master/pkg/cloudprovider/provider/gce/config.go#L165

**Review by ahmedwaleedmalik — CHANGES_REQUESTED** [review]

File: pkg/cloudprovider/provider/gce/config.go (hunk 3 above)
Comment (tlamr) [reply pkg/cloudprovider/provider/gce/config.go:105]: would it work to have some sensible defaults?

File: pkg/cloudprovider/provider/gce/provider.go (hunk 2 above)
Comment (tlamr) [reply pkg/cloudprovider/provider/gce/provider.go:270]: what if we always provide an ok default?

File: pkg/cloudprovider/provider/gce/provider.go (hunk 4 above)
Comment (tlamr) [reply pkg/cloudprovider/provider/gce/provider.go:269]: You are totally right, sorry for that!

File: pkg/cloudprovider/provider/gce/provider.go (hunk 2 above)
Comment (ahmedwaleedmalik) [reply pkg/cloudprovider/provider/gce/provider.go:270]: It's already going to be defaulted in the upstream API(google) if you don't provide a value. In that case I think it makes complete sense to use pointers and just rely on their defaults.

File: pkg/cloudprovider/provider/gce/config.go (hunk 3 above)
Comment (ahmedwaleedmalik) [reply pkg/cloudprovider/provider/gce/config.go:105]: It's already going to be defaulted in the upstream API(google) if you don't provide a value. In that case I think it makes complete sense to use pointers and just rely on their defaults.

Comment (tlamr) [reply pkg/cloudprovider/provider/gce/config.go]: ok, did my best, hopefully should be ok now. Would using "" as empty value for string work in go? Using pointer there would mean I'm unable to write pass it nicely in that struct the gcp uses.

Thanks @ahmedwaleedmalik for all the patience with my beginner go issues :)

Comment (ahmedwaleedmalik) [reply pkg/cloudprovider/provider/gce/config.go]: > Thanks @ahmedwaleedmalik for all the patience with my beginner go issues :)

Happy to help ✌️

**Comment by ahmedwaleedmalik** [comment]
@tlamr Thank you for the awesome work, it's a very valuable addition to the GCE config.

I have made some amendments with https://github.com/kubermatic/machine-controller/pull/1252/commits/1383ada7c828c12a4be73d3168edbc4c40d516e6. Can you please take a look? This is what I was trying to suggest earlier. Do let me know if that makes sense for you.

**Review by ahmedwaleedmalik — APPROVED** [review]
/approve 🚀

**Comment by tlamr** [comment]
> 1383ada

ah, ok, thanks a lot! :)

**Pull request merged**